src/parser/simplify.rs

```diff
@@ -25,57 +25,88 @@ use crate::target::Target;
 use crate::document_tree::{
 	Document,
 	HasChildren,
-	attribute_types::ID,
+	attribute_types::{ID, NameToken},
 	elements as e,
 	element_categories as c,
 };
 
 
-enum MaybeDirectTarget {
-	IndirectTarget(ID),
-	DirectTarget(Target),
+#[derive(Debug)]
+enum NamedTargetType {
+	NumberedFootnote(usize),
+	LabeledFootnote(usize),
+	Citation,
+	InternalLink,
+	ExternalLink(Target),
+	IndirectLink(NameToken),
+	SectionTitle
+}
+impl NamedTargetType {
+	fn is_implicit_target(&self) -> bool {
+		match self {
+			NamedTargetType::SectionTitle => true,
+			_ => false
+		}
+	}
+}
+
+#[derive(Clone, Debug)]
+struct Substitution {
+	content: e::ImageInline,
+	// If `ltrim` is true and the sibling before the reference is a text node,
+	// the text node gets right-trimmed. Same for `rtrim` with the sibling after
+	// getting left-trimmed.
+	ltrim: bool,
+	rtrim: bool
```

Comment:
Are those only useful here or also in the tree?

Reply:
Those would affect the nodes around a substitution reference after it's substituted (see the TODO in line 366). It's yet another feature that can't quite be dealt with `resolve_refs`, which considers only the current node and can't modify its surroundings. I just forgot that this was a thing when drafting the PR message/